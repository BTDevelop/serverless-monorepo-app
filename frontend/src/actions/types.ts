const _REQUESTED = '_REQUESTED';
const _FULFILLED = '_FULFILLED';
const _REJECTED = '_REJECTED';

export const LOGIN_REQUESTED = 'LOGIN' + _REQUESTED;
export const LOGIN_FULFILLED = 'LOGIN' + _FULFILLED;
export const LOGIN_REJECTED = 'LOGIN' + _REJECTED;

export const LOGOUT_REQUESTED = 'LOGOUT' + _REQUESTED;
export const LOGOUT_FULFILLED = 'LOGOUT' + _FULFILLED;
export const LOGOUT_REJECTED = 'LOGOUT' + _REJECTED;

export const SYNC_AUTH_STATE_REQUESTED = 'SYNC_AUTH_STATE' + _REQUESTED;
export const SYNC_AUTH_STATE_FULFILLED = 'SYNC_AUTH_STATE' + _FULFILLED;
export const SYNC_AUTH_STATE_REJECTED = 'SYNC_AUTH_STATE' + _REJECTED;

export const API_REQUESTED = 'API' + _REQUESTED;
export const API_FULFILLED = 'API' + _FULFILLED;
export const API_REJECTED = 'API' + _REJECTED;
