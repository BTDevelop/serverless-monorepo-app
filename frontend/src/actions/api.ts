import { createAction } from 'redux-actions';
import * as types from './types';

export const apiMetaTypes = {
  dbCreate: 'dbCreate',
  email: 'email',
  file: 'file',
  private: 'private',
  public: 'public',
};

export const apiRequested = createAction(
  types.API_REQUESTED,
  undefined,
  (_, type) => ({ type }),
);
export const apiFulfilled = createAction(
  types.API_FULFILLED,
  undefined,
  (_, type) => ({ type }),
);
export const apiRejected = createAction(
  types.API_REJECTED,
  undefined,
  (_, type) => ({ type }),
);

export const emailApiRequested = (toAddress: string) =>
  apiRequested(toAddress, apiMetaTypes.email);

export const privateApiRequested = () =>
  apiRequested(undefined, apiMetaTypes.private);

export const publicApiRequested = () =>
  apiRequested(undefined, apiMetaTypes.public);

export const fileApiRequested = (fileUrl: string, key: string) =>
  apiRequested({ fileUrl, key }, apiMetaTypes.file);

export const dbCreateApiRequested = (text: string) =>
  apiRequested(text, apiMetaTypes.dbCreate);
