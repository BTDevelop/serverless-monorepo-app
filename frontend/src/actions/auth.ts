import { createAction } from 'redux-actions';
import * as types from './types';

export const loginRequested = createAction(types.LOGIN_REQUESTED);
export const loginFulfilled = createAction(types.LOGIN_FULFILLED);
export const loginRejected = createAction(types.LOGIN_REJECTED);

export const logoutRequested = createAction(types.LOGOUT_REQUESTED);
export const logoutFulfilled = createAction(types.LOGOUT_FULFILLED);
export const logoutRejected = createAction(types.LOGOUT_REJECTED);

export const syncAuthStateRequested = createAction(
  types.SYNC_AUTH_STATE_REQUESTED,
);
export const syncAuthStateFulfilled = createAction(
  types.SYNC_AUTH_STATE_FULFILLED,
);
export const syncAuthStateRejected = createAction(
  types.SYNC_AUTH_STATE_REJECTED,
);
