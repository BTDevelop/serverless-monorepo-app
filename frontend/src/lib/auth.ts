import { Auth0DecodedHash, WebAuth } from 'auth0-js';

const auth = new WebAuth({
  audience: process.env.REACT_APP_AUTH0_AUDIENCE,
  clientID: process.env.REACT_APP_AUTH0_CLIENT_ID || '',
  domain: process.env.REACT_APP_AUTH0_DOMAIM || '',
  redirectUri: process.env.REACT_APP_AUTH0_REDIRECT_URI,
  responseType: 'token id_token',
  scope: 'openid profile email',
});

const setSession = (authResult: Auth0DecodedHash) => {
  const { expiresIn = 0, accessToken = '', idToken = '' } = authResult;
  // Set the time that the access token will expire at
  const expiresAt = JSON.stringify(expiresIn * 1000 + new Date().getTime());
  localStorage.setItem('access_token', accessToken);
  localStorage.setItem('id_token', idToken);
  localStorage.setItem('expires_at', expiresAt);
};

const handleAuthentication = async () => {
  const promise: Promise<Auth0DecodedHash> = new Promise((resolve, reject) => {
    auth.parseHash((err, authResult) => {
      const result = authResult || {};
      if (err) {
        reject(err);
      } else {
        setSession(result);
        resolve(result);
      }
    });
  });
  return await promise;
};

const login = (state: string) => auth.authorize({ state });

const logout = () => {
  localStorage.removeItem('access_token');
  localStorage.removeItem('id_token');
  localStorage.removeItem('expires_at');
};

const isAuthenticated = () => {
  // Check whether the current time is past the
  // access token's expiry time
  const expiresAt = JSON.parse(
    localStorage.getItem('expires_at') || JSON.stringify(0),
  );
  return new Date().getTime() < expiresAt;
};

export { handleAuthentication, login, logout, isAuthenticated };
