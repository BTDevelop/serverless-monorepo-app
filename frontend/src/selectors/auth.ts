import { IState } from '../types/redux';

export const isAuthenticated = (state: IState) => state.auth.authenticated;
export const isNotAuthenticated = (state: IState) => !state.auth.authenticated;
export const isInitialized = (state: IState) => state.auth.initialized;
