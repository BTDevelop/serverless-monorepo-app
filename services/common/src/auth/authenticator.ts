import jwt = require('jsonwebtoken');

// Set in `environment` of serverless.yml
const AUTH0_CLIENT_ID = process.env.AUTH0_CLIENT_ID;
// istanbul ignore next
const AUTH0_CLIENT_PUBLIC_KEY = process.env.AUTH0_CLIENT_PUBLIC_KEY || '';

// Policy helper function
export const generatePolicy = (
  principalId: string,
  effect: string,
  resource: string,
) => {
  if (effect && resource) {
    const policyDocument = {
      Statement: [
        {
          Action: 'execute-api:Invoke',
          Effect: effect,
          Resource: resource,
        },
      ],
      Version: '2012-10-17',
    };
    return { principalId, policyDocument };
  } else {
    return { principalId };
  }
};

export const decodeToken = async (
  authorizationToken: string | undefined,
  methodArn: string,
) => {
  const error = 'Unauthorized';
  if (!authorizationToken) {
    console.log('missing authorizationToken');
    return { response: null, error };
  }

  const tokenParts = authorizationToken.split(' ');
  const tokenValue = tokenParts[1];

  if (!(tokenParts[0].toLowerCase() === 'bearer' && tokenValue)) {
    console.log('invalid authorizationToken value');
    return { response: null, error };
  }

  const options = {
    audience: AUTH0_CLIENT_ID,
  };

  const promise = new Promise((resolve, reject) => {
    jwt.verify(tokenValue, AUTH0_CLIENT_PUBLIC_KEY, options, (err, decoded) => {
      if (err) {
        reject(err);
      } else {
        resolve(decoded);
      }
    });
  });

  try {
    const decoded = (await promise) as { sub: string };
    console.log('valid from customAuthorizer', decoded);
    const response = exports.generatePolicy(decoded.sub, 'Allow', methodArn);
    return { response, error: null };
  } catch (err) {
    console.log('verifyError', err);
    return { response: null, error };
  }
};
