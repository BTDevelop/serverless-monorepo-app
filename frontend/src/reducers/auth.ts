import { Action, handleActions } from 'redux-actions';
import * as types from '../actions/types';
import { IAuthState } from '../types/redux';

const defaultState: IAuthState = {
  authenticated: false,
  error: null,
  inProgress: false,
  initialized: false,
};

const reducer = handleActions<IAuthState, any>(
  {
    [types.LOGIN_REQUESTED]: state => {
      return { ...state, inProgress: true };
    },
    [types.LOGIN_FULFILLED]: state => {
      return { ...state, inProgress: false, authenticated: true, error: null };
    },
    [types.LOGIN_REJECTED]: (state, action) => {
      return {
        ...state,
        authenticated: false,
        error: action.payload,
        inProgress: false,
      };
    },
    [types.LOGOUT_REQUESTED]: state => {
      return { ...state, inProgress: true };
    },
    [types.LOGOUT_FULFILLED]: state => {
      return { ...state, inProgress: false, authenticated: false, error: null };
    },
    [types.LOGOUT_REJECTED]: (state, action: Action<Error>) => {
      return {
        ...state,
        authenticated: false,
        error: action.payload,
        inProgress: false,
      };
    },
    [types.SYNC_AUTH_STATE_REQUESTED]: state => {
      return { ...state, inProgress: true };
    },
    [types.SYNC_AUTH_STATE_FULFILLED]: (_, action: Action<boolean>) => {
      return {
        authenticated: !!action.payload,
        error: null,
        inProgress: false,
        initialized: true,
      };
    },
    [types.SYNC_AUTH_STATE_REJECTED]: (_, action: Action<Error>) => {
      return {
        authenticated: false,
        error: action.payload,
        inProgress: false,
        initialized: true,
      };
    },
  },
  defaultState,
);

export default reducer;
