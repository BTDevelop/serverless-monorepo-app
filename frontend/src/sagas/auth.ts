import { call, put, select, takeEvery } from 'redux-saga/effects';
import * as types from '../actions/types';
import { isAuthenticated, login, logout } from '../lib/auth';
import {
  logoutFulfilled,
  logoutRejected,
  syncAuthStateFulfilled,
  syncAuthStateRejected,
} from './../actions/auth';
import { redirectRouteSelector } from './../selectors/router';

export function* handleLogin() {
  const redirectRoute = yield select(redirectRouteSelector);
  yield call(login, redirectRoute);
}

export function* loginSaga() {
  yield takeEvery(types.LOGIN_REQUESTED, handleLogin);
}

export function* handleLogout() {
  try {
    yield call(logout);
    yield put(logoutFulfilled());
  } catch (err) {
    yield put(logoutRejected(err));
  }
}

export function* logoutSaga() {
  yield takeEvery(types.LOGOUT_REQUESTED, handleLogout);
}

export function* handleAuthState() {
  try {
    const authenticated = yield call(isAuthenticated);
    yield put(syncAuthStateFulfilled(authenticated));
  } catch (err) {
    yield put(syncAuthStateRejected(err));
  }
}

export function* authStateSaga() {
  yield takeEvery(types.SYNC_AUTH_STATE_REQUESTED, handleAuthState);
}

const sagas = [loginSaga(), logoutSaga(), authStateSaga()];

export default sagas;
