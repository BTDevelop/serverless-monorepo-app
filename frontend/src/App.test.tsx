import { shallow } from 'enzyme';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router-dom';
import * as renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { syncAuthStateRequested as syncAuthState } from './actions/auth';
import ConnectedApp, { App } from './App';

const middlewares: any[] = [];
const mockStore = configureStore(middlewares);

jest.mock('./components/Api', () => {
  return {
    default: 'api-component',
  };
});

jest.mock('./components/AuthStatus', () => {
  return {
    default: 'auth-status-component',
  };
});

jest.mock('./components/Loading', () => {
  return {
    ComponentLoading: 'component-loading-component',
    default: 'loading-component',
  };
});

jest.mock('./selectors/auth');
jest.mock('./selectors/router');

const props = {
  initialized: true,
  pathname: '/public',
  syncAuthState: jest.fn(),
};

describe('App', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <App {...props} />
      </MemoryRouter>,
      div,
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  test('renders correctly initialized=true', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <App {...props} />
        </MemoryRouter>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders correctly initialized=false', () => {
    const tree = renderer
      .create(<App {...props} initialized={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('syncs auth state on mount', () => {
    const wrapper = shallow(<App {...props} />, {
      disableLifecycleMethods: true,
    });

    // @ts-ignore access to undefined
    wrapper.instance().componentDidMount();

    expect(props.syncAuthState).toHaveBeenCalledTimes(1);
  });

  test('should connect component correctly', () => {
    const { isInitialized } = require('./selectors/auth');
    isInitialized.mockReturnValueOnce(true);

    const { pathnameSelector } = require('./selectors/router');
    pathnameSelector.mockReturnValueOnce('somepath');

    const state = { someState: 'someState' };
    const store = mockStore(state);

    // @ts-ignore wrong store type
    const wrapper = shallow(<ConnectedApp store={store} />);

    expect(wrapper.props().initialized).toBe(true);
    expect(wrapper.props().pathname).toBe('somepath');
    wrapper.props().syncAuthState();

    expect(store.getActions()).toHaveLength(1);
    expect(store.getActions()[0]).toEqual(syncAuthState());

    expect(isInitialized).toHaveBeenCalledTimes(1);
    expect(isInitialized).toHaveBeenCalledWith(state);

    expect(pathnameSelector).toHaveBeenCalledTimes(1);
    expect(pathnameSelector).toHaveBeenCalledWith(state);
  });
});
