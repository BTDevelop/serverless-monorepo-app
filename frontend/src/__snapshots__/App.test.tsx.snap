// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly initialized=false 1`] = `
<div
  className="ui container"
>
  <loading-component />
</div>
`;

exports[`App renders correctly initialized=true 1`] = `
<div
  className="ui container"
>
  <div
    style={
      Object {
        "margin": "0 auto",
        "width": 1000,
      }
    }
  >
    <div
      className="ui segment"
    >
      <ul>
        <li>
          <a
            href="/public"
            onClick={[Function]}
          >
             Public Page 
          </a>
        </li>
        <li>
          <a
            href="/private"
            onClick={[Function]}
          >
             Private Private 
          </a>
        </li>
      </ul>
    </div>
    <div
      className="ui segment"
    >
      Welcome!
    </div>
    <auth-status-component />
    <api-component />
  </div>
</div>
`;
